src/main/scala/chisel3/util/Counter.scala

```diff
@@ -24,22 +24,45 @@ import chisel3.internal.naming.chiselName  // can't use chisel3_ version because
   * maximum output value of the counter), need not be a power of two
   */
 @chiselName
-class Counter(val n: Int) {
-  require(n >= 0, s"Counter value must be nonnegative, got: $n")
-  val value = if (n > 1) RegInit(0.U(log2Ceil(n).W)) else 0.U
+class Counter private (r: Range) {
+  require(r.start >= 0 && r.end >= 0, s"Counter range must be positive, got: $r")
 
-  /** Increment the counter
+  private val delta = math.abs(r.step).U
+  private val width = math.max(log2Up(r.start), log2Up(r.end)) + 1
+
+  /** Creates a counter with the specified number of steps.
+    *
+    * @param n number of steps before the counter resets
+    */
+  def this(n: Int) { this(0 until n) }
+
+  /** The current value of the counter. */
+  val value = if (r.length > 1) RegInit(r.start.U(width.W)) else r.start.U
+
+  /** The range of the counter values. */
+  def range: Range = r
+
+  /** Increments the counter by a step.
     *
     * @note The incremented value is registered and will be visible on the next clock cycle
-    * @return whether the counter will wrap to zero on the next cycle
+    * @return whether the counter will wrap on the next clock cycle
     */
   def inc(): Bool = {
-    if (n > 1) {
-      val wrap = value === (n-1).U
-      value := value + 1.U
-      if (!isPow2(n)) {
```

Comment:
I'm not sure that this `isPow2` check can still be used, as a range allows stepping by numbers other than one.

Reply:
I think the new implementation here more than covers that case.

My only [mild] concern is that power-of-2 size Counters that start at 0 may be marginally more efficient with the old specialized logic taking advantage of rollover. There is a slight difference in the emitted Verilog:

Before:
```verilog
  reg [3:0] count; // @[Counter.scala 29:33]
  wire [3:0] _T_2 = count + 4'h1; // @[Counter.scala 39:22]
  always @(posedge clock) begin
    if (reset) begin
      count <= 4'h0;
    end else if (io_en) begin
      count <= _T_2;  // Hopefully we will be inlining this sooner rather than latter
    end
```

After:
```verilog
  reg [4:0] count; // @[Counter.scala 37:40]
  wire  _T = count == 5'hf; // @[Counter.scala 49:24]
  wire [4:0] _T_2 = count + 5'h1; // @[Counter.scala 56:26]
  always @(posedge clock) begin
    if (reset) begin
      count <= 5'h0;
    end else if (io_en) begin
      if (_T) begin
        count <= 5'h0;
      end else begin
        count <= _T_2;
      end
    end
  end
```

I think we should eliminate that extra `if`. It may be that synthesis tools can figure this out, but even if they can it'd be ideal to keep the Verilog for that common case the same. While I don't like the style, it might be simplest just to have special logic for that case.

Also the width is too big.

Reply:
Yosys, at least, doesn't optimize the mux.  Keep the special case.

Reply:
I have re-added the optimization to let the counter overflow naturally when it is possible.